crates/amaru-consensus/src/consensus/tx_submission/tx_submission_client_state.rs

```diff
@@ -0,0 +1,170 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use amaru_kernel::peer::Peer;
+use amaru_ouroboros_traits::{MempoolSeqNo, TxId, TxServerRequest, TxSubmissionMempool};
+use anyhow::anyhow;
+use serde::{Deserialize, Serialize};
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::sync::Arc;
+use tracing::debug;
+
+/// State of a transaction submission client for a given peer.
+///
+/// We keep track of:
+///
+///  - Which transaction ids have been advertised to the peer but not yet fully acknowledged.
+///  - The last sequence number we pulled from the mempool for this peer, so we know where to continue from.
+///
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TxSubmissionClientState {
+    /// Peer we are serving.
+    peer: Peer,
+    /// What we’ve already advertised but has not yet been fully acked.
+    window: VecDeque<(TxId, MempoolSeqNo)>,
+    /// Last seq_no we have ever pulled from the mempool for this peer.
+    /// None if we have not pulled anything yet.
+    last_seq: Option<MempoolSeqNo>,
+}
+
+impl TxSubmissionClientState {
+    pub fn new(peer: &Peer) -> Self {
+        Self {
+            peer: peer.clone(),
+            window: VecDeque::new(),
+            last_seq: None,
+        }
+    }
+
+    pub fn peer(&self) -> &Peer {
+        &self.peer
+    }
+
+    /// Process a request from the tx submission server
+    pub async fn process_tx_server_request<Tx: Send + Debug + Sync + 'static>(
+        &mut self,
+        mempool: Arc<dyn TxSubmissionMempool<Tx>>,
+        request: TxServerRequest,
+    ) -> anyhow::Result<TxClientResponse<Tx>> {
+        match request {
+            TxServerRequest::TxIds { ack, req, .. } => {
+                if req == 0 {
+                    debug!(peer = %self.peer,
+                        "Requested 0 tx ids, terminating tx submission client",
+                    );
+                    return Ok(TxClientResponse::<Tx>::ProtocolError(anyhow!(
+                        "0 transaction ids requested"
+                    )));
+                }
+                // update the window by discarding acknowledged tx ids
+                // and update the last_seq
+                self.discard(ack);
+                if !mempool
+                    .wait_for_at_least(self.last_seq.unwrap_or_default().add(req as u64))
+                    .await
+                {
+                    return Ok(TxClientResponse::<Tx>::Done);
+                }
+                let tx_ids = self.get_next_tx_ids(mempool, req).await?;
+                Ok(TxClientResponse::NextIds(tx_ids))
+            }
+            TxServerRequest::TxIdsNonBlocking { ack, req, .. } => {
```

Comment:
according to the spec we need to check that exactly one of `ack` and `req` is non-zero (which sounds stupid to me because I can well imagine setting both to positive values)